demos/applications/apa102.cpp

```diff
@@ -57,66 +57,71 @@ void application(resource_list& p_map)
-  hal::display::apa102_pixel blue{ .blue = 0xFF, .green = 0x00, .red = 0x00 };
-  hal::display::apa102_pixel white{ .blue = 0xFF, .green = 0xFF, .red = 0xFF };
+  // predefined color arrays
   std::array<hal::display::apa102_pixel, led_count> all_off{};
-
-  std::array<hal::display::apa102_pixel, led_count> rgb_array = {
-    { white, blue, green, red }
+  std::array<hal::display::apa102_pixel, 4> predefined_colors = {
+    { { .blue = 0xFF, .green = 0xFF, .red = 0xFF },
+      { .blue = 0xFF, .green = 0x00, .red = 0x00 },
+      { .blue = 0x00, .green = 0xFF, .red = 0x00 },
+      { .blue = 0x00, .green = 0x00, .red = 0xFF } }
   };
 
+  std::array<hal::display::apa102_pixel, led_count> rgb_array;
+  for (uint8_t i = 0; i < led_count; i++) {
+    rgb_array[i] = predefined_colors[i % 4];
+  }
+
   hal::print(console, "Demo Application Starting...\n\n");
   hal::display::apa102 led_strip(spi, chip_select);
   while (true) {
     // reset LEDs by turning them all off
     update_all(all_off, brightness, apa_frame);
 
+    hal::print(console, "Updating single LEDS\n");
     // update one at a time, all other LEDs should remain the same state
     // added delays to visually see individual activations
-    hal::print(console, "Updating single LEDS\n");
-    update_single(red, brightness, 0, apa_frame);
-    led_strip.update(apa_frame);
-    hal::delay(clock, 500ms);
-    update_single(green, brightness, 1, apa_frame);
-    led_strip.update(apa_frame);
-    hal::delay(clock, 500ms);
-    update_single(blue, brightness, 2, apa_frame);
-    led_strip.update(apa_frame);
-    hal::delay(clock, 500ms);
-    update_single(white, brightness, 3, apa_frame);
-    led_strip.update(apa_frame);
+    for (uint8_t i = 0; i < led_count; i++) {
+      update_single(predefined_colors[i % 4], brightness, i, apa_frame);
+      led_strip.update(apa_frame);
+      hal::delay(clock, 500ms);
+    }
+    hal::delay(clock, 3s);
+
+    // reset LEDs by turning them all off
+    update_all(all_off, brightness, apa_frame);
 
     // update all LEDs at once
-    hal::delay(clock, 3s);
     hal::print(console, "Updating all LEDS\n");
     update_all(rgb_array, brightness, apa_frame);
     led_strip.update(apa_frame);
 
     // cycle through RGB colors, start and end with red
-    hal::delay(clock, 3s);
-    hal::display::apa102_pixel rainbow = red;
     hal::print(console, "Rainbow Cycle\n");
+    hal::delay(clock, 3s);
+    hal::display::apa102_pixel rainbow = { .blue = 0xFF,
+                                           .green = 0xFF,
+                                           .red = 0xFF };
+    std::array<hal::display::apa102_pixel, led_count> rainbow_array;
+
     for (int i = 0; i <= 255; i++) {
       // decrease red, increase blue
       rainbow.red = 255 - i;
       rainbow.blue = i;
-      std::array<hal::display::apa102_pixel, led_count> rainbow_array = {
-        { rainbow, rainbow, rainbow, rainbow }
-      };
+      for (uint8_t x = 0; x < led_count; x++) {
```

Comment:
```suggestion
      for (decltype(led_count) x = 0; x < led_count; x++) {
```

`decltype` captures the type of led_count and uses it. Since you defined `led_count` as unsigned, it could be set to a value above 255 which is the max for `uint8_t`. So its better to make sure these match.